### PR TITLE
[RFC] Making nvim big-endian friendly

### PR DESCRIPTION
**Pull request by jamessan**
Since I have easy access to big-endian systems via Debian's porterboxes, I'm
working through the big-endian specific issues being caught by the [builds].

[builds]: https://buildd.debian.org/neovim

### CODE REVIEW

**Comment by bfredl** [comment]
@jamessan Do you plan look into autoread as well? (If so, feel free to steal the test from the latest commit of #3917 if you find it useful.)

File: src/nvim/api/window.c [hunk 1]
```diff
@@ -348,15 +348,15 @@ Tabpage nvim_win_get_tabpage(Window window, Error *err)
 /// @return Window number
 Integer nvim_win_get_number(Window window, Error *err)
 {
-  Integer rv = 0;
+  handle_T rv = 0;
```
Comment (bfredl) [inline src/nvim/api/window.c]: window numbers are not handles, so this should just be `int`

File: src/nvim/api/private/helpers.c [hunk 2]
```diff
@@ -615,15 +615,34 @@ bool object_to_vim(Object obj, typval_T *tv, Error *err)
     case kObjectTypeBuffer:
     case kObjectTypeWindow:
     case kObjectTypeTabpage:
-    case kObjectTypeInteger:
-      if (obj.data.integer > INT_MAX || obj.data.integer < INT_MIN) {
+    case kObjectTypeInteger: {
```
Comment (bfredl) [inline src/nvim/api/private/helpers.c]: can't we just always store the handle as int64 in the `obj.data` union and then add casts as necessary? (most cases are already be handled by `gendispatch.lua`)

Comment (jamessan) [inline src/nvim/api/private/helpers.c]: That seems more invasive than just making this change.

**Comment by jamessan** [comment]
> Do you plan look into autoread as well?

Making it behave like Vim does?  Isn't that a separate issue?

**Comment by bfredl** [comment]
>  Isn't that a separate issue?

Sure. I just saw you mention it on gitter, so I wanted to make you aware there was a test. But if not I'll be happy look into it again myself. (I kind of forgot about #3917 because there was seemingly no interest in fixing big-endian support at the time)

**Comment by jamessan** [comment]
> I just saw you mention it on gitter, so I wanted to make you aware there was a test.

Yeah, that was when I was still considering changing everything to `bool`.  I didn't realize that in the process of defaulting `'autoread'` to on, the behavior of how the value is propagated had deviated from Vim.

File: src/nvim/api/private/helpers.c (hunk 2 above)
Comment (bfredl) [inline src/nvim/api/private/helpers.c]: Yes, I think that would be the best. (One could consider `int64_t buffer` but that would be confusing with the type difference.)

**Comment by jamessan** [comment]
Updated to use `obj.data.integer` for buffer, window, and tabpage.

**Comment by justinmk** [comment]
LGTM

File: src/nvim/api/private/helpers.c [hunk 3]
```diff
@@ -615,14 +615,24 @@ bool object_to_vim(Object obj, typval_T *tv, Error *err)
     case kObjectTypeBuffer:
     case kObjectTypeWindow:
     case kObjectTypeTabpage:
+      if (obj.data.integer > INT_MAX || obj.data.integer < 0) {
```
Comment (bfredl) [inline src/nvim/api/private/helpers.c]: I don't think checking that handles are nonnegative warrants having a separate switch arm. If checked, this should be checked at Object construction, here it is too late. Conversion should check that it fits the target domain, that is VARNUMBER_MIN` to `VARNUMBER_MAX`.

Comment (jamessan) [inline src/nvim/api/private/helpers.c]: Fixed.

**Comment by bfredl** [comment]
LGTM

**Comment by justinmk** [comment]
I think the QB fail is a random QB quirk, it happened on another PR. Might want to force-push just to be sure.

**Comment by jamessan** [comment]
Taking another spin on the QB wheel.

**Comment by jamessan** [comment]
Yay, everything's green.

**Pull request merged**